src/coreclr/jit/llvm_types.h

```diff
@@ -15,6 +15,7 @@ struct TypeDescriptor
 private:
     unsigned              m_fieldCount;
     CORINFO_FIELD_HANDLE* m_fields;
+    unsigned*             m_fieldSizes;
```

Comment:
How are the field sized here different from those you get with `getClassSize`? Is the difference that they include alignment? This needs to be documented.

Reply:
I don't think they are, the problem is how to get the field class when in the JIT.  I did it to reduce the interop partly and also you could ask for the class size from the `CORINFO_FIELD_HANDLE` but we'd need to add a callback for it as there is nothing in the interface.

Reply:
> the problem is how to get the field class when in the JIT

Wouldn't `getFieldType` + `getClassSize` work here?

I think it is ok to take the interop cost hit for this building code, as it is run only once per type per process. Holding the (enlarged) type descriptor data structures in memory is not desirable (we want the type descriptor itself be as small as reasonable, as it is supposed to only be used for interop, and we already duplicate "the type system" on the on the Jit side with the `StructDesc*` map (that is supposed to be for "hot" data).

Another way to approach this is to build the `StructDesc*`s for field descriptors greedily, i. e. before filtering out the union-ed fields, and then save and get the size there.

Reply:
I dont think so, `getFieldType` does not return the `CORINFO_CLASS_STRUCT_` for the field.
```
        private CorInfoType getFieldType(CORINFO_FIELD_STRUCT_* field, CORINFO_CLASS_STRUCT_** structType, CORINFO_CLASS_STRUCT_* memberParent)
        {
            FieldDesc fieldDesc = HandleToObject(field);
            TypeDesc fieldType = fieldDesc.FieldType;

            CorInfoType type;
            if (structType != null)
            {
                type = asCorInfoType(fieldType, structType);
            }
            else
            {
                type = asCorInfoType(fieldType);
            }

            Debug.Assert(!fieldDesc.OwningType.IsByReferenceOfT ||
                fieldDesc.OwningType.GetKnownField("_value").FieldType.Category == TypeFlags.IntPtr);
            if (type == CorInfoType.CORINFO_TYPE_NATIVEINT && fieldDesc.OwningType.IsByReferenceOfT)
            {
                Debug.Assert(structType == null || *structType == null);
                Debug.Assert(fieldDesc.Offset.AsInt == 0);
                type = CorInfoType.CORINFO_TYPE_BYREF;
            }

            return type;
        }
```

Reply:
Isn't that what the `CORINFO_CLASS_STRUCT_** structType` out parameter does?

Reply:
missed that, will try.